File: src/main/java/de/flavormate/ba_entities/category/repository/CategoryRepository.java
package de.flavormate.ba_entities.category.repository;

import de.flavormate.ba_entities.category.model.Category;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;
import java.util.Optional;

public interface CategoryRepository extends JpaRepository<Category, Long> {
	@Query("select c from Category c join c.localizations l where lower(l.value) = lower(:label) order by l.value")
	Optional<Category> findByLabel(@Param("label") String label);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = ?1 order by l.value")
	Page<Category> findByPage(String language, Pageable pageable);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = ?1 order by l.value")
	List<Category> findAll(String language);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = ?1 and c.id = ?2 order by l.value")
	List<Category> findById(String language, Long id);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = ?1 and c.recipes is not empty order by l.value")
	Page<Category> findByNotEmpty(String language, Pageable pageable);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = :language and lower(l.value) like lower(concat('%', :search, '%')) order by l.value")
	Page<Category> findBySearch(@Param("language") String language, @Param("search") String searchTerm, Pageable pageable);

	@Query("select new de.flavormate.ba_entities.category.model.Category(c, l.value) from Category c join c.localizations l where l.id.language = :language order by l.value")
	List<Category> findByRaw(String language);
}

File: src/main/kotlin/de/flavormate/features/category/repositories/CategoryRepository.kt
/* Licensed under AGPLv3 2024 - 2025 */
package de.flavormate.features.category.repositories

import de.flavormate.features.category.daos.models.CategoryEntity
import de.flavormate.features.recipe.daos.models.RecipeEntity
import io.quarkus.hibernate.orm.panache.kotlin.PanacheQuery
import io.quarkus.hibernate.orm.panache.kotlin.PanacheRepositoryBase
import io.quarkus.panache.common.Sort
import jakarta.enterprise.context.ApplicationScoped

@ApplicationScoped
class CategoryRepository : PanacheRepositoryBase<CategoryEntity, String> {

  override fun findAll(sort: Sort): PanacheQuery<CategoryEntity> {
      return find(
          query = "select c from CategoryEntity c left join  c.localizations l",
      sort = sort,
    )
  }

    fun findAll(sort: Sort, language: String): PanacheQuery<CategoryEntity> {
        val params = mapOf("language" to language)
        return find(
            query = "select c from CategoryEntity c left join c.localizations l where l.id.language = :language and c.recipes is not empty",
            sort = sort,
            params = params,
        )
  }

    fun findByLocalizedLabelAndLanguage(label: String): CategoryEntity? {
        val params = mapOf("label" to label)
    return find(
        "select c from CategoryEntity c left join c.localizations l where lower(l.value) = lower(:label)",
        params,
      )
      .firstResult()
  }

  fun findBySearch(sort: Sort, language: String, query: String): PanacheQuery<CategoryEntity> {
    val params = mapOf("language" to language, "query" to query)
    return find(
        "select c from CategoryEntity c left join c.localizations l where l.id.language = :language and lower(l.value) like lower(concat('%', :query, '%'))",
      sort,
      params,
    )
  }

  fun countRecipes(id: String): PanacheQuery<Long> {
    val params = mapOf("id" to id)

    return find("select count(r) from CategoryEntity c join c.recipes r where c.id = :id", params)
      .project(Long::class.java)
  }

  fun findRecipes(sort: Sort, id: String): PanacheQuery<RecipeEntity> {
    val params = mapOf("id" to id)

    return find("select r from CategoryEntity c join c.recipes r where c.id = :id", sort, params)
      .project(RecipeEntity::class.java)
  }

  fun findByIdsIn(ids: List<String>): List<CategoryEntity> {
    val params = mapOf("ids" to ids)

    return list(query = "id in :ids", params = params)
  }
}
